compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -373,11 +413,108 @@ impl<'tcx> SizeSkeleton<'tcx> {
             (SizeSkeleton::Pointer { tail: a, .. }, SizeSkeleton::Pointer { tail: b, .. }) => {
                 a == b
             }
+            // constants are always pre-normalized into a canonical form so this
+            // only needs to check if their pointers are identical.
+            (SizeSkeleton::Generic(a), SizeSkeleton::Generic(b)) => a == b,
             _ => false,
         }
     }
 }
 
+/// To compare constants, we only use this to multiply them, which ensures
+/// that each multiplied term is in sorted order.
+/// i.e. b * a always gets normalized to a * b, in order of the Ord impl
+/// of `Const`.
+/// It will return `None` if multiplication of constants overflows.
+fn mul_sorted_consts<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    a: ty::Const<'tcx>,
+    b: ty::Const<'tcx>,
+) -> Option<ty::Const<'tcx>> {
+    use crate::mir::BinOp::Mul;
+    use ty::ConstKind::Expr;
+    use ty::Expr::Binop;
+
+    // fast-path
+    match (a.kind(), b.kind()) {
+        (Expr(Binop(Mul, ..)), _) | (_, Expr(Binop(Mul, ..))) => {}
+        _ => {
+            let (l, r) = if a < b { (a, b) } else { (b, a) };
+            return Some(tcx.mk_const(Expr(Binop(Mul, l, r)), tcx.types.usize));
+        }
+    }
+
+    let mut buf: SmallVec<[ty::Const<'tcx>; 4]> = SmallVec::new();
+    let mut a = sort_mul(tcx, param_env, a)?;
+    let mut b = sort_mul(tcx, param_env, b)?;
+    while let (Expr(Binop(Mul, al, ar)), Expr(Binop(Mul, bl, br))) = (a.kind(), b.kind()) {
+        if al < bl {
+            buf.push(al);
+            a = ar;
+        } else {
+            buf.push(bl);
+            b = br;
+        }
+    }
+
+    while let Expr(Binop(Mul, al, ar)) = a.kind() {
+        if al < b {
+            buf.push(al);
+        } else {
+            buf.push(std::mem::replace(&mut b, al));
+        }
+        a = ar;
+    }
+    while let Expr(Binop(Mul, bl, br)) = b.kind() {
+        if bl < a {
+            buf.push(bl);
+        } else {
+            buf.push(std::mem::replace(&mut a, bl));
+        }
+        b = br;
+    }
+
+    if a < b {
+        buf.push(a);
+        buf.push(b);
+    } else {
+        buf.push(b);
+        buf.push(a);
+    }
+
+    // buf is now sorted minimum -> maximum, with evaluatable at the end.
```

Comment:
This is not true, take e.g. `fn transpose<const W: usize, const H: usize>(v: [[u32; 2 * H]; W + W]) -> [[u32; W + W]; 2 * H]`.

Here in the `compute` call for `[[u32; 2 * H]; W + W]` we get the literal `8` because this statement holds for `[u32; 2*H]`, so we get `Mul(H, Mul(8, Add(W, W)))` for the whole term, whereas for `[[u32; W + W]; 2 * H]` we get `Mul(4, Add(W, W))` for the inner term `[u32; W + W]`, but then we can't simplify to the literal `8` because `Literal(n) < Binop(Add, W, W)`, so we end up with `Mul(H, Mul(2, Mul(4, Add(W, W))))`. 

I think you would have to either manually implement `Ord` or write a function that orders this solely for this functionality.